File: python/metatag/import_io/itunes_sync.py
"""iTunes library import/export."""

import logging
import plistlib
import shutil
from pathlib import Path
from typing import List
from urllib.parse import quote, unquote, urlparse

from ..core.track import Track

logger = logging.getLogger(__name__)


def itunes_url_to_path(url: str) -> Path:
    """Convert iTunes file:// URL to local Path."""
    parsed = urlparse(url)
    if parsed.scheme != "file":
        raise ValueError(f"Not a file URL: {url}")

    if parsed.netloc:
        path = f"//{parsed.netloc}{parsed.path}"
    else:
        path = parsed.path

    # Windows: file:///C:/Users/... → /C:/Users/… → C:/Users/…
    if path.startswith("/") and len(path) > 2 and path[2] == ":":
        path = path[1:]

    return Path(unquote(path))


def path_to_itunes_url(path: Path) -> str:
    """Convert a local Path back to the file:// URL format iTunes uses.

    iTunes uses percent-encoded file:// URIs.  On Windows the path becomes
    file:///C:/Users/… (three slashes, forward-slashes, percent-encoded).
    """
    # Normalise to forward slashes
    posix_str = str(path).replace("\\", "/")
    if not posix_str.startswith("/"):
        # Windows drive letter — add leading slash
        posix_str = "/" + posix_str
    # Percent-encode everything except safe chars
    return "file://" + quote(posix_str, safe="/:@!$&'()*+,;=")


def import_library(xml_path: str) -> List[Track]:
    """Import tracks from iTunes library XML."""
    tracks = []
    try:
        with open(xml_path, "rb") as f:
            plist = plistlib.load(f)
    except Exception as e:
        logger.error(f"Failed to read iTunes library plist: {e}")
        return []

    tracks_dict = plist.get("Tracks")
    if not isinstance(tracks_dict, dict):
        logger.error("No 'Tracks' dictionary in iTunes library")
        return []

    for _track_id, track_info in tracks_dict.items():
        if not isinstance(track_info, dict):
            continue

        location = track_info.get("Location")
        if not location:
            continue

        try:
            file_path = itunes_url_to_path(location)
        except ValueError as e:
            logger.warning(f"Invalid file URL {location}: {e}")
            continue

        if not file_path.exists():
            logger.warning(f"File does not exist: {file_path}")
            continue

        track = Track(file_path)
        if not track.load():
            logger.warning(f"Failed to load track: {file_path}")
            continue

        # Override with iTunes metadata where present
        _map_itunes_fields(track_info, track)
        tracks.append(track)

    logger.info(f"Imported {len(tracks)} tracks from iTunes library")
    return tracks


def _map_itunes_fields(track_info: dict, track: Track) -> None:
    """Apply iTunes plist field dict to a Track object."""
    if artist := track_info.get("Artist"):
        track.artist = str(artist)
    if album := track_info.get("Album"):
        track.album = str(album)
    if title := track_info.get("Name"):
        track.title = str(title)
    if genre := track_info.get("Genre"):
        track.genre = str(genre)
    if composer := track_info.get("Composer"):
        track.composer = str(composer)
    if grouping := track_info.get("Grouping"):
        track.grouping = str(grouping)
    if comment := track_info.get("Comments"):
        track.comment = str(comment)
    if isinstance(n := track_info.get("Track Number"), int) and n > 0:
        track.track_number = n
    if isinstance(d := track_info.get("Disc Number"), int) and d > 0:
        track.disc_number = d
    if isinstance(y := track_info.get("Year"), int) and y > 0:
        track.year = y
    if isinstance(b := track_info.get("BPM"), int) and b > 0:
        track.bpm = b


def export_changes(tracks: List[Track], xml_path: str) -> bool:
    """Export track metadata changes back to an iTunes Library XML file.

    The function:
    1. Reads the existing iTunes XML.
    2. Builds a mapping of file:// URL → Tracks entry.
    3. Updates every matching entry with current Track metadata.
    4. Creates a .bak backup of the original, then writes the updated XML.

    Returns True on success, False on any failure.
    """
    xml_file = Path(xml_path)
    if not xml_file.exists():
        logger.error(f"iTunes XML not found: {xml_path}")
        return False

    # ── Load ──────────────────────────────────────────────────────────────
    try:
        with open(xml_file, "rb") as fh:
            plist = plistlib.load(fh)
    except Exception as e:
        logger.error(f"Failed to read iTunes library: {e}")
        return False

    tracks_dict = plist.get("Tracks")
    if not isinstance(tracks_dict, dict):
        logger.error("Malformed iTunes XML: missing 'Tracks' dictionary")
        return False

    # ── Build URL→entry map ───────────────────────────────────────────────
    # Key = normalised file URL string; value = the mutable plist sub-dict
    url_to_entry: dict[str, dict] = {}
    for entry in tracks_dict.values():
        if isinstance(entry, dict) and (loc := entry.get("Location")):
            # Normalise to lower-case for case-insensitive path matching
            url_to_entry[loc.lower()] = entry

    updated = 0
    for track in tracks:
        url = path_to_itunes_url(track.file_path)
        entry = url_to_entry.get(url.lower())
        if entry is None:
            logger.debug(f"Track not found in iTunes XML: {track.file_path.name}")
            continue

        # Write back every field that iTunes supports
        entry["Name"]    = track.title
        entry["Artist"]  = track.artist
        entry["Album"]   = track.album
        entry["Genre"]   = track.genre
        if track.composer:
            entry["Composer"] = track.composer
        if track.grouping:
            entry["Grouping"] = track.grouping
        if track.comment:
            entry["Comments"] = track.comment
        if track.track_number > 0:
            entry["Track Number"] = track.track_number
        if track.track_total > 0:
            entry["Track Count"] = track.track_total
        if track.disc_number > 0:
            entry["Disc Number"] = track.disc_number
        if track.disc_total > 0:
            entry["Disc Count"] = track.disc_total
        if track.year > 0:
            entry["Year"] = track.year
        if track.bpm > 0:
            entry["BPM"] = track.bpm

        updated += 1

    if updated == 0:
        logger.warning("No tracks matched in iTunes XML — nothing to export")
        return False

    # ── Backup original ───────────────────────────────────────────────────
    backup = xml_file.with_suffix(".bak")
    try:
        shutil.copy2(xml_file, backup)
        logger.info(f"iTunes XML backup saved to {backup}")
    except Exception as e:
        logger.warning(f"Could not create backup: {e}")

    # ── Write updated XML ─────────────────────────────────────────────────
    try:
        with open(xml_file, "wb") as fh:
            plistlib.dump(plist, fh, fmt=plistlib.FMT_XML)
        logger.info(f"Exported changes for {updated} track(s) to {xml_path}")
        return True
    except Exception as e:
        logger.error(f"Failed to write iTunes XML: {e}")
        return False

File: src/import/itunes_sync.cpp
#include "itunes_sync.h"
#include "track.h"
#include <QFile>
#include <QXmlStreamReader>
#include <QDebug>
#include <QUrl>

namespace metatag {

static bool parseTracksDict(QXmlStreamReader& xml, QVector<Track>& tracks);

bool iTunesSync::importLibrary(const QString& xmlPath, QVector<Track>& tracks)
{
    QFile file(xmlPath);
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qWarning() << "Failed to open iTunes library XML:" << xmlPath;
        return false;
    }
    
    QXmlStreamReader xml(&file);
    
    while (!xml.atEnd() && !xml.hasError()) {
        xml.readNext();
        if (xml.isStartElement() && xml.name() == QStringLiteral("plist")) {
            // Read plist contents
            while (!xml.atEnd() && !xml.hasError()) {
                xml.readNext();
                if (xml.isEndElement() && xml.name() == QStringLiteral("plist"))
                    break;
                if (xml.isStartElement() && xml.name() == QStringLiteral("dict")) {
                    // Parse top-level dict
                    while (!xml.atEnd() && !xml.hasError()) {
                        xml.readNext();
                        if (xml.isEndElement() && xml.name() == QStringLiteral("dict"))
                            break;
                        if (xml.isStartElement() && xml.name() == QStringLiteral("key")) {
                            QString key = xml.readElementText();
                            if (key == QStringLiteral("Tracks")) {
                                xml.readNext(); // start of dict
                                if (xml.isStartElement() && xml.name() == QStringLiteral("dict")) {
                                    if (!parseTracksDict(xml, tracks))
                                        return false;
                                }
                            } else {
                                // Skip value
                                xml.readNext();
                                if (xml.isStartElement() && xml.name() == QStringLiteral("dict")) {
                                    // skip entire dict
                                    int depth = 1;
                                    while (depth > 0 && !xml.atEnd() && !xml.hasError()) {
                                        xml.readNext();
                                        if (xml.isStartElement())
                                            depth++;
                                        else if (xml.isEndElement())
                                            depth--;
                                    }
                                } else if (xml.isStartElement()) {
                                    xml.skipCurrentElement();
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    
    if (xml.hasError()) {
        qWarning() << "XML error:" << xml.errorString();
        return false;
    }
    
    file.close();
    return true;
}

static bool parseTracksDict(QXmlStreamReader& xml, QVector<Track>& tracks)
{
    while (!xml.atEnd() && !xml.hasError()) {
        xml.readNext();
        if (xml.isEndElement() && xml.name() == QStringLiteral("dict"))
            break;
        
        if (xml.isStartElement() && xml.name() == QStringLiteral("key")) {
            QString trackId = xml.readElementText(); // track ID, ignore
            xml.readNext(); // start of dict
            if (xml.isStartElement() && xml.name() == QStringLiteral("dict")) {
                // parse track properties
                QString location;
                QString artist, album, title, genre, composer, grouping, comment;
                int trackNumber = 0, discNumber = 0, year = 0, bpm = 0;
                
                while (!xml.atEnd() && !xml.hasError()) {
                    xml.readNext();
                    if (xml.isEndElement() && xml.name() == QStringLiteral("dict"))
                        break;
                    if (xml.isStartElement() && xml.name() == QStringLiteral("key")) {
                        QString key = xml.readElementText();
                        xml.readNext(); // value
                        if (xml.isCharacters() || xml.isStartElement()) {
                            QString value;
                            if (xml.isCharacters())
                                value = xml.text().toString();
                            else if (xml.isStartElement() && xml.name() == QStringLiteral("string"))
                                value = xml.readElementText();
                            
                            if (key == QStringLiteral("Location"))
                                location = QUrl(value).toLocalFile();
                            else if (key == QStringLiteral("Artist"))
                                artist = value;
                            else if (key == QStringLiteral("Album"))
                                album = value;
                            else if (key == QStringLiteral("Name"))
                                title = value;
                            else if (key == QStringLiteral("Genre"))
                                genre = value;
                            else if (key == QStringLiteral("Composer"))
                                composer = value;
                            else if (key == QStringLiteral("Grouping"))
                                grouping = value;
                            else if (key == QStringLiteral("Comments"))
                                comment = value;
                            else if (key == QStringLiteral("Track Number"))
                                trackNumber = value.toInt();
                            else if (key == QStringLiteral("Disc Number"))
                                discNumber = value.toInt();
                            else if (key == QStringLiteral("Year"))
                                year = value.toInt();
                            else if (key == QStringLiteral("BPM"))
                                bpm = value.toInt();
                        }
                        // skip remaining elements
                        if (xml.isStartElement())
                            xml.skipCurrentElement();
                    }
                }
                
                if (!location.isEmpty()) {
                    // Convert location to local path (remove URL encoding and file://)
                    QString filePath = QUrl(location).toLocalFile();
                    if (!filePath.isEmpty()) {
                        Track track(filePath);
                        if (track.load()) {
                            if (!artist.isEmpty())
                                track.setArtist(artist);
                            if (!album.isEmpty())
                                track.setAlbum(album);
                            if (!title.isEmpty())
                                track.setTitle(title);
                            if (!genre.isEmpty())
                                track.setGenre(genre);
                            if (!composer.isEmpty())
                                track.setComposer(composer);
                            if (!grouping.isEmpty())
                                track.setGrouping(grouping);
                            if (!comment.isEmpty())
                                track.setComment(comment);
                            if (trackNumber > 0)
                                track.setTrackNumber(trackNumber);
                            if (discNumber > 0)
                                track.setDiscNumber(discNumber);
                            if (year > 0)
                                track.setYear(year);
                            if (bpm > 0)
                                track.setBpm(bpm);
                            
                            tracks.append(track);
                        }
                    }
                }
            }
        }
    }
    
    return !xml.hasError();
}

bool iTunesSync::exportChanges(const QVector<Track>& tracks)
{
    // For now, just log that this is not implemented
    // A full implementation would need to:
    // 1. Read the existing iTunes XML
    // 2. Update track entries with new metadata
    // 3. Write back the XML
    // This is complex and error-prone, so we'll leave it unimplemented for now

    qWarning() << "iTunes export is not implemented. Changes were not exported to iTunes library.";
    return false;
}

}